src/comm/KernelDiagnostics.cc

```diff
@@ -0,0 +1,60 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file KernelDiagnostics.cc
+//---------------------------------------------------------------------------//
+#include "KernelDiagnostics.hh"
+
+#include <iostream>
+#include "base/Macros.hh"
+#include "base/Range.hh"
+#include "comm/Device.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Global reference to shared Celeritas kernel diagnostics.
```

Comment:
Since (at the moment?) KernelDiagnostics is not thread safe, this implies that we are planning to have all kernel interactions are done by a single/main threads, right?

Reply:
Also a good point. When we're doing most of the work on GPU as opposed to CPU, it may be OK to avoid shared memory parallelism on the host. Regardless though, these global functions definitely need comments/todos marking the lack of thread safety. What thread models do you have experience with? The CPU code in Shift can use OpenMP for shared-memory parallelism.

Reply:
Generally I would not envision ever calling kernel diagnostics from within a host side thread parallel region.  Even if we ran a MPI+OpenMP+CUDA model, my expectation would be that we would always attach 1GPU per MPI rank, and each MPI rank could hypothetically execute thread-parallel regions on multicore on the host - side, but those parallel regions would always be outside the kernel launches.  In other words, I don't foresee a use-case where kernels are launched from individual threads.

And yes @pcanal I know it "can" be done, but it doesn't mean its a good idea ;)

Reply:
I guess @pcanal is coming from a standpoint where CPU parallelism is usually done through threads a la geant?

Reply:
Yes, thread or (worse for this case) thread indirectly via tasking model (TBB for example).  (For the record Geant4 is migrating (slowly) from the earlier to the later.

We will (eventually) need to interface with the experiment framework which use the tasking model.  They can mark the simulation module as "single-threaded" if need be but it is better for general concurrency if it is not.  An option is for the simulation to "appear" multi-thread where it can take input from multiple thread/task at once but process them internally single threaded. [In this context the "input" provided by the experiment framework is the primary particles (instead of getting them from HepMC or a file).

Reply:
> And yes @pcanal I know it "can" be done, but it doesn't mean its a good idea ;)

Maybe (or maybe not) it makes sense to have one thread per actual device.

Reply:
Yeah there's definitely some exploration needed there -- does it make sense to require exactly one GPU per Geant4 thread? I would imagine each geant thread's incident track would act as a separate "primary" farmed out to the GPU, and we'd have some sort of threadsafe queueing system to batch up tracks to send to the device. But for "now" for the initial testing purposes I think the geant/celeritas coupling can be single-threaded.

I added a line in the device initialization with a link to these conversations and an assertion if OpenMP and CUDA are enabled simultaneously.

Reply:
Yes, the devil is in the details and in actual measurements :).

So yes, I agree for this phase we can do whatever is more convenience/more performant and some point soon-ish (when we have some inkling that we have good enough results/prospects) we will need to start the conversation with users to work out the interface and flows ...